src/sinks/tcp.rs

```diff
@@ -61,23 +128,93 @@ impl SinkConfig for TcpSinkConfig {
     }
 }
 
+fn load_certificate<T: AsRef<Path> + Debug>(filename: T) -> Result<Certificate, String> {
+    open_read_parse(filename, "certificate authority", Certificate::from_pem)
+}
+
+fn load_key<T: AsRef<Path> + Debug>(
+    filename: T,
+    pass_phrase: &Option<String>,
+) -> Result<PKey<Private>, String> {
+    match pass_phrase {
+        None => open_read_parse(filename, "key", PKey::private_key_from_pem),
+        Some(phrase) => open_read_parse(filename, "key", |data| {
+            PKey::private_key_from_pem_passphrase(data, phrase.as_bytes())
+        }),
+    }
+}
+
+fn load_x509<T: AsRef<Path> + Debug>(filename: T) -> Result<X509, String> {
+    open_read_parse(filename, "certificate", X509::from_pem)
+}
+
+fn open_read_parse<F: AsRef<Path> + Debug, O, E: Error, P: Fn(&[u8]) -> Result<O, E>>(
+    filename: F,
+    note: &str,
+    parser: P,
+) -> Result<O, String> {
+    let mut text = Vec::<u8>::new();
+
+    File::open(filename.as_ref())
+        .map_err(|err| format!("Could not open {} file {:?}: {}", note, filename, err))?
+        .read_to_end(&mut text)
+        .map_err(|err| format!("Could not read {} file {:?}: {}", note, filename, err))?;
+
+    parser(&text).map_err(|err| format!("Could not parse {} file {:?}: {}", note, filename, err))
+}
+
 pub struct TcpSink {
+    hostname: String,
     addr: SocketAddr,
+    tls: Option<TcpSinkTls>,
     state: TcpSinkState,
     backoff: ExponentialBackoff,
 }
 
 enum TcpSinkState {
     Disconnected,
     Connecting(ConnectFuture),
-    Connected(FramedWrite<TcpStream, BytesCodec>),
+    TlsConnecting(TlsConnect<TcpStream>),
+    Connected(Box<dyn FramedConnection + Send>),
```

Comment:
We can avoid the box here by using something like a `MaybeTlsStream` like so https://github.com/sfackler/rust-postgres/blob/c87e58a31cc48b4e427e6df7db99dabdb79c97b1/tokio-postgres/src/proto/maybe_tls_stream.rs#L6

though this isn't a blocker right now just a suggestion

Reply:
Sure, can do. The box is cheap, but it turns out the match pattern is faster.